File: src/main/java/com/github/morningzeng/toolset/component/RadioBar.java
package com.github.morningzeng.toolset.component;

import com.intellij.ui.components.JBPanel;
import com.intellij.ui.components.JBPanelWithEmptyText;
import com.intellij.ui.components.JBRadioButton;

import javax.swing.ButtonGroup;
import java.awt.event.ItemEvent;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;

/**
 * @author Morning Zeng
 * @since 2024-07-17
 */
public final class RadioBar<T> extends JBPanel<JBPanelWithEmptyText> {

    private final Map<T, JBRadioButton> radioMap;
    private final ButtonGroup group;

    @SafeVarargs
    public RadioBar(final T defaultCheck, final T... radios) {
        this(defaultCheck, Arrays.asList(radios));
    }

    public RadioBar(final T defaultCheck, final List<T> radios) {
        this.group = new ButtonGroup();
        this.radioMap = radios.stream().collect(
                Collectors.toMap(Function.identity(), item -> {
                    final JBRadioButton radio = new JBRadioButton(item.toString(), item.equals(defaultCheck));
                    this.add(radio);
                    this.group.add(radio);
                    return radio;
                })
        );
        this.radioMap.forEach((t, radio) -> radio.addItemListener(e -> {
            if (e.getStateChange() == ItemEvent.SELECTED) {
                radioMap.values().stream()
                        .filter(Predicate.not(radio::equals))
                        .forEach(item -> item.setSelected(false));
            }
        }));
    }

    public void addItemListener(final Consumer<? super T> action) {
        this.radioMap.keySet().forEach(t -> this.addItemListener(t, action));
    }

    public void addItemListener(final T t, final Consumer<? super T> action) {
        this.radioMap.get(t).addItemListener(e -> {
            if (e.getStateChange() == ItemEvent.SELECTED) {
                action.accept(t);
            }
        });
    }

    public void addItemListener(final Consumer<? super T> action, Predicate<T> predicate) {
        this.radioMap.keySet().stream()
                .filter(predicate)
                .forEach(t -> this.addItemListener(t, action));
    }

    @Override
    public void setEnabled(final boolean enabled) {
        super.setEnabled(enabled);
        this.radioMap.values().forEach(radio -> radio.setEnabled(enabled));
    }
}

File: src/main/kotlin/com/github/morningzeng/toolbox/ui/bar/RadioBar.kt
@file:Suppress("unused")

package com.github.morningzeng.toolbox.ui.bar

import com.intellij.ui.components.JBPanel
import com.intellij.ui.components.JBPanelWithEmptyText
import com.intellij.ui.components.JBRadioButton
import java.awt.event.ItemEvent
import javax.swing.BoxLayout
import javax.swing.ButtonGroup

/**
 * @author Morning Zeng
 * @since 2025-07-23
 */
class RadioBar<T>(vertical: Boolean, selected: T, items: List<T>) : JBPanel<JBPanelWithEmptyText>() {

    private val radioMap: Map<T, JBRadioButton>
    private val group: ButtonGroup = ButtonGroup()
    private var selected: T? = selected
    private val changeListeners: MutableList<(T?) -> Unit> = mutableListOf()

    constructor(vararg items: T) : this(false, items[0], items.toList())

    constructor(vertical: Boolean, selected: T, vararg items: T) : this(vertical, selected, items.toList())

    constructor(items: List<T>) : this(false, items[0], items.toList())

    constructor(vertical: Boolean, vararg items: T) : this(vertical, items[0], items.toList())

    constructor(selected: T, items: List<T>) : this(false, selected, items)

    init {
        this.layout = BoxLayout(this, if (vertical) BoxLayout.Y_AXIS else BoxLayout.X_AXIS)
        radioMap = items.associateWith {
            JBRadioButton(it.toString(), it == selected).also { xit ->
                add(xit)
                group.add(xit)
                xit.addActionListener { cit ->
                    fireChangeEvent(it)
                }
            }
        }
        radioMap.forEach { (t, radio) ->
            radio.addItemListener {
                if (it.stateChange == ItemEvent.SELECTED) {
                    setSelected(t)
                }
            }
        }
    }

    fun getSelected(): T? {
        return selected
    }

    fun setSelected(selected: T?) {
        if (selected == this.selected) return
        this.selected?.let { radioMap[it]?.isSelected = false }
        radioMap[selected]?.also {
            it.isSelected = true
            this.selected = selected
        }
        fireChangeEvent(selected)
    }

    fun addChangeListener(listener: (T?) -> Unit) {
        changeListeners.add(listener)
    }

    fun removeChangeListener(listener: (T?) -> Unit) {
        changeListeners.remove(listener)
    }

    private fun fireChangeEvent(selected: T?) {
        changeListeners.forEach {
            try {
                it(selected)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun setEnabled(enabled: Boolean) {
        super.setEnabled(enabled)
        radioMap.values.forEach { it.isEnabled = enabled }
    }

}